gravitee-am-policy/gravitee-am-policy-enroll-mfa/src/main/java/io/gravitee/am/policy/enroll/mfa/EnrollMfaPolicy.java

```diff
@@ -0,0 +1,183 @@
+                            __ -> {
+                                LOGGER.debug("MFA factor with ID [{}] enrolled for user {}", factorId, user.getId());
+                                policyChain.doNext(request, response);
+                            },
+                            error -> {
+                                LOGGER.error("Unable to enroll MFA factor with ID [{}]", factorId, error.getMessage());
+                                policyChain.failWith(PolicyResult.failure(GATEWAY_POLICY_ENROLL_MFA_ERROR_KEY, error.getMessage()));
+                            });
+
+
+        } catch (Exception ex) {
+            LOGGER.error("An error has occurred for [enroll-mfa] policy", ex);
+            policyChain.failWith(PolicyResult.failure(GATEWAY_POLICY_ENROLL_MFA_ERROR_KEY, ex.getMessage()));
+        }
+    }
+
+    private Single<EnrolledFactor> buildEnrolledFactor(Factor factor, User user, String value, ExecutionContext context) {
+        return Single.defer(() -> {
+            try {
+                // compute value
+                String enrollmentValue = (!ObjectUtils.isEmpty(value)) ? context.getTemplateEngine().getValue(value, String.class) : null;
+                // create the enrolled factor
+                EnrolledFactor enrolledFactor = new EnrolledFactor();
+                enrolledFactor.setFactorId(factor.getId());
+                enrolledFactor.setStatus(FactorStatus.PENDING_ACTIVATION);
+                enrolledFactor.setPrimary(configuration.isPrimary());
+                switch (factor.getFactorType()) {
+                    case OTP:
+                        enrolledFactor.setSecurity(new EnrolledFactorSecurity(SHARED_SECRET, enrollmentValue));
+                        break;
+                    case SMS:
+                        enrolledFactor.setChannel(new EnrolledFactorChannel(EnrolledFactorChannel.Type.SMS, enrollmentValue));
+                        break;
+                    case CALL:
+                        enrolledFactor.setChannel(new EnrolledFactorChannel(EnrolledFactorChannel.Type.CALL, enrollmentValue));
+                        break;
+                    case EMAIL:
+                        Map<String, Object> additionalData = Collections.singletonMap(FactorDataKeys.KEY_MOVING_FACTOR, generateInitialMovingFactor(user));
+                        enrolledFactor.setSecurity(new EnrolledFactorSecurity(SHARED_SECRET, SharedSecret.generate(), additionalData));
+                        enrolledFactor.setChannel(new EnrolledFactorChannel(EnrolledFactorChannel.Type.EMAIL, enrollmentValue));
+                        break;
+                    case HTTP:
+                        enrolledFactor.setChannel(new EnrolledFactorChannel(EnrolledFactorChannel.Type.HTTP, enrollmentValue));
+                        break;
+                    default:
+                        return Single.error(new IllegalStateException("Unexpected value: " + factor.getFactorType().getType()));
```

Comment:
When the content of the EL is `null` we return an `Internal Server Error`.
Is it up to the user to make sure the expression language returns something ? Or should we redirect to the Enroll page to make sure the user isn't blocked ? We could just let the user continue through the flow and if he hasn't enroll he will presented with the Enroll page ?

WDYT 🤔  ?

![image](https://user-images.githubusercontent.com/8531515/171000204-0f550fe5-feb3-4aa8-b83b-c59566197242.png)

Reply:
FP